### DISCUSSION TITLE
Retry on Migration Failure running in a Transaction

### DISCUSSION THREAD

**Discussion by StephBaldacchino** [flamingock/mongock#61] [Ideas]
### Description

Mongock provides support for native MongoDB transactions as well as Spring Data transactions. Would be nice to have Mongock to also support retry on failed transactions. This would provide the option that in case of a failed transaction, the transaction would still be rolled back (as it currently does) but additionally, it could be re-tried from scratch automatically. This could be achieved by using Spring Retry. This would be very practical since MongoDB do not lock the respective documents during migration, but rather just fails and rollbacks if there is a conflicting change while the migration is running.

### Specific scenario

In our current application, we’re able to run DB migrations on application startup by using Mongock. We have also recently started benefiting from having transactions enabled (across all migration) when using Mongock. This was required for High Availability setup. In line with this, we came across the scenario where in case of a failed migration running in a transaction, it might be wishful and quite convenient to have subsequent attempts/retries to run the migration again.

### PRIORITY

Minor

### Expected behavior

Having some mongock property which defines transaction-retry; on failed transaction, by default no retries are made (current solution), but in case of this property having positive custom value, further attempts are made to run the migration.

### PRIORITY

NICE_TO_HAVE

### Additional Information

In our HA setup, a failed migration running in a transaction might occur quite frequently since while one component is ongoing startup and running new migration on a collection, an older component instance might be changing some documents in the same collection. This would result in failure but in reality an auto-retry would be ideal.

**Comment by dieppa** [comment]
Hello, first thanks for your contribution, @StephBaldacchino 

I think that the current solution is fine when the application can afford a stop-the-world approach, specially having some kind of orchestrator tool, like Kubernetes, which restarts the service(implicitly retries the migration).

However in the scenario you describe it's a bit cumbersome. I think it's good to have that kind of mechanism you suggest, we'll add it to our road map, thanks! :bowtie: 

PD: As we have already started the development for version 5, we are not adding more features for version 4, just bug fixes.
Our expectation for the release of version 5 is around May.
I see it's not a priority for your team. If that changes and you need it, please send us an email to dev@cloudyrock.io, so we can try to find a solution for you.

PD 2: Regardless that I see the value of your suggestion, and I don't have enough information about your system/architecture, the approach you provide is a bit risky. You should be pay special attention to services accessing to obsolete structure and so on.
For migrations in HA services, I recommend a 2-step migration. First one, a backward compatible migration where both kind of instances, the ones with the updated structure and the ones with old structure, can co-live together. Remove the old ones, and then the second step, where I can safely run a migration to remove the obsolete structure.

You probably have taken this into account already, but just in case 😉  

PD 3: Can I ask for which company you are using Mongock?

Regards

**Reply by StephBaldacchino** [reply]
Hello, thanks for your feedback.

Indeed, the stop-the-world approach is not so bad provided tools exists which could restarting the service automatically, which is why I've set priority as minor since it would be a nice to have feature. However, an application startup failure might have been caused by anything and might not always be ideal to just opt for an auto-restart.

As you've suggested, we're handling migrations in HA environment in the described '2-step migration'. But thanks for bringing it up. :smiley: 

Thanks for consider the feature, appreciated.

**Reply by dieppa** [reply]
I am glad you are ok with the answer and we'll definitely add this feature tu the next version.

I don't know if you forgot to provide the name of your company or you can't/want. That's fine, could I ask the industry then? This kind of information is quite useful for us.

Thanks

**Reply by StephBaldacchino** [reply]
I work for phoenixNAP and we've been using Mongock for the last 2 years since version 3.0.0-BETA of the library. We've found it really useful. So, great work :slightly_smiling_face:

phoenixNAP is an IT infrastructure provider offering cloud, hosting and bare-metal services.
